django/template/base.py

```diff
@@ -314,38 +476,43 @@ def parse(self, parse_until=None):
                 self.extend_nodelist(nodelist, TextNode(token.contents), token)
             elif token.token_type == 1:  # TOKEN_VAR
                 if not token.contents:
-                    self.empty_variable(token)
+                    raise self.error(token, 'Empty variable tag')
                 try:
                     filter_expression = self.compile_filter(token.contents)
                 except TemplateSyntaxError as e:
-                    if not self.compile_filter_error(token, e):
-                        raise
-                var_node = self.create_variable_node(filter_expression)
+                    raise self.error(token, e)
+                var_node = VariableNode(filter_expression)
                 self.extend_nodelist(nodelist, var_node, token)
             elif token.token_type == 2:  # TOKEN_BLOCK
                 try:
                     command = token.contents.split()[0]
                 except IndexError:
-                    self.empty_block_tag(token)
+                    raise self.error(token, 'Empty block tag')
                 if command in parse_until:
-                    # put token back on token list so calling
-                    # code knows why it terminated
+                    # A matching token has been reached. Return control to
+                    # the caller. Put the token back on the token list so the
+                    # caller knows where it terminated.
                     self.prepend_token(token)
                     return nodelist
-                # execute callback function for this tag and append
-                # resulting node
-                self.enter_command(command, token)
+                # Add the token to the command stack. This is used for error
+                # messages if further parsing fails due to an unclosed block
+                # tag.
+                self.command_stack.append((command, token))
+                # Get the tag callback function from the ones registered with
+                # the parser.
                 try:
                     compile_func = self.tags[command]
                 except KeyError:
                     self.invalid_block_tag(token, command, parse_until)
+                # Compile the callback into a node object and add it to
+                # the node list.
                 try:
                     compiled_result = compile_func(self, token)
-                except TemplateSyntaxError as e:
-                    if not self.compile_function_error(token, e):
-                        raise
+                except Exception as e:
```

Comment:
I changed this from ``TemplateSyntaxError`` to ``Exception``. See the test I added: ``template_tests.tests.TemplateTests.test_compile_tag_error``.